pkgs/by-name/ka/kak-tree-sitter/package.nix

```diff
@@ -0,0 +1,48 @@
+{
+  lib,
+  rustPlatform,
+  fetchFromSourcehut,
+  makeWrapper,
+  tinycc,
+  nix-update-script,
+  testers,
+  kak-tree-sitter,
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "kak-tree-sitter";
+  version = "1.1.2";
+
+  src = fetchFromSourcehut {
+    owner = "~hadronized";
+    repo = "kak-tree-sitter";
+    rev = "kak-tree-sitter-v${version}";
+    hash = "sha256-wBWfSyR8LGtug/mCD0bJ4lbdN3trIA/03AnCxZoEOSA=";
+  };
+
+  cargoHash = "sha256-OQPUWqJAts8DbFNSsC/CmMCbuZ9TVxRTR05O7oiodKI=";
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  # Tree-Sitter grammars are C programs that need to be compiled
```

Comment:
I know few to nothing about tree-sitter and lsp servers.
How does it work? 
Are C source files compiled when the editor starts, or something like that? Why can't these files be precompiled?

Reply:
> How does it work?

Tree sitter grammars are written using a javascript library that generates efficient C code. 
That C code is normally committed in the same repository as the original javascript code.
This weird process makes it easy for developers to write new grammars, and easy for code editors to install new grammars, all you need is a C compiler:
https://github.com/helix-editor/helix/blob/ff33b07756548935577aefc15cf48a1beb27b162/languages.toml#L512-L514

> Are C source files compiled when the editor starts, or something like that?

Something like that, but only once of course, for kak-tree-sitter you can install all declared grammars using `ktsctl sync --all`. If you add another grammar to your config, you can run it again.

> Why can't these files be precompiled?

They certainly can! Many are already packaged in `nixpkgs` [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/tools/parsing/tree-sitter/grammars).

I'm already using them, but it's janky for now: https://git.lelgenio.com/lelgenio/nixos-config/src/commit/9da25c99fe2f4e2255c52f6cdb65b2b54f98e119/user/kakoune/kak-tree-sitter.nix

The problem is that you'll frequently find yourself needing to install some grammar that is not packaged in `nixpkgs` or just don't want nix consume your life :laughing:

Adding `cc` to the `PATH` of `ktsctl` was not mandatory for this package to be useful, but since the consensus on `nixpkgs` is to add optional dependencies by default, I went with that.

Reply:
Hum, I am a bit skeptical about bringing a C compiler to the scope of a package.
To me it looks more natural to let this decision to the user.

Thinking in the perspective of a user of another distro like Ubuntu or Rhino, theey would install two packages: the C compiler and the kak-tree-sitter.
Maybe a smarter user would create a metapackage for declarative package installation - https://ayats.org/blog/declarative-ubuntu - but this is a whole another topic.

Point is, a typical user is expected to install both packages (plus some book-keeping like the wrapper you did).

Now, back to Nixpkgs.

Another problem is that we are creating an artificial dependency link here: ~if~ when tcc is updated, kak-tree-sitter will be needlessly recompiled (given that kak-tree-sitter does not use tcc as a real dependency upstream).

That being said, a better solution would be creating a third package, say `kak-tree-sitter-plus-tcc`, basically a wrapper that merges both packages in a `symlinkJoin`.
I did something like this in x16-emulator, remember me later.